### PR TITLE
lvmde code

### PR DESCRIPTION
**Pull request by PierreBoyeau**

### CODE REVIEW

File: scvi/model/base/_demixin.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+        n_samples_loop = z_reshaped.shape[0]
+        log_px_zs = []
+        for _tensors in scdl:
+            inference_inputs = self.module._get_inference_input(_tensors)
+            n_cells = inference_inputs["batch_index"].shape[0]
+            zs = z_reshaped.expand(n_samples_loop, n_cells, self.module.n_latent)
+            log_px_zs.append(
+                self.module.estimate_likelihood(
+                    tensors=_tensors, z=zs, library=None
+                ).cpu()
```
Comment (adamgayoso) [inline scvi/model/base/_demixin.py:290]: why not use lossoutput?

File: scvi/model/base/_demixin.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+            h = generative_outputs["px"].scale
+            n_genes = h.shape[-1]
+            h = h.reshape(-1, n_genes).cpu()
+
+            zs.append(inference_outputs["z"].reshape(-1, self.module.n_latent).cpu())
+            qzs_m.append(inference_outputs["qz"].loc.cpu())
+            qzs_std.append(inference_outputs["qz"].scale.cpu())
+            hs.append(h)
+
+            _log_px_zs = self.evaluate_likelihood_from_z(scdl, inference_outputs["z"])
```
Comment (adamgayoso) [inline scvi/model/base/_demixin.py:213]: this is running through the decoder twice, why not just compute the loss and use the lossoutput?

File: scvi/module/_vae.py [hunk 3]
```diff
@@ -518,16 +553,20 @@ def sample(
 
     @torch.inference_mode()
     @auto_move_data
-    def marginal_ll(self, tensors, n_mc_samples):
+    def marginal_ll(
+        self, tensors, n_mc_samples, observation_specific=False, n_mc_samples_per_pass=1
```
Comment (adamgayoso) [inline scvi/module/_vae.py:557]: n_mc_samples_per_pass is like chunking, we need to validate that the chunking is less than n_mc_samples?

File: scvi/model/base/_rnamixin.py [hunk 4]
```diff
@@ -181,14 +181,13 @@ def differential_expression(
         batchid2: Optional[Iterable[str]] = None,
         fdr_target: float = 0.05,
         silent: bool = False,
+        importance_sampling: Optional[bool] = False,
```
[the remark this answers is not in the document]
Comment (PierreBoyeau) [reply scvi/model/base/_rnamixin.py:184]: This makes sense. My main fear was that `RNAMixin` would then be too large, but it should be ok. I am on it!

File: scvi/model/base/_demixin.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+            indices=indices,
+            give_mean=True,
+            batch_size=batch_size,
+        )
+        if (qz_m.ndim != 2) or (qz_m.shape[0] != len(indices)):
+            raise ValueError("Dimension mismatch of variational density means")
+        try:
+            idx_filt = EllipticEnvelope().fit_predict(qz_m)
+            idx_filt = idx_filt == 1
+        except ValueError:
```
[the remark this answers is not in the document]
Comment (PierreBoyeau) [reply scvi/model/base/_demixin.py:321]: I think this piece is important. Sometimes the `EllipticEnvelope` fails and we need to run back to something else. I am going to see if I can more easily identify which cases raise `ValueError`s~

File: scvi/model/base/_demixin.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+            idx_filt = np.ones(qz_m.shape[0], dtype=bool)
+        try:
+            logger.debug(
+                "n cells total {}, n cells filtered {}".format(
+                    idx_filt.shape[0], idx_filt.sum()
+                )
+            )
+            indices = indices[idx_filt]
+        except IndexError:
+            raise IndexError(idx_filt)
```
[the remark this answers is not in the document]
Comment (PierreBoyeau) [reply scvi/model/base/_demixin.py:334]: I agree, let me remove it.

File: scvi/model/base/_rnamixin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/model/base/_rnamixin.py:184]: wait!, it should be the other way around @PierreBoyeau

**Comment by adamgayoso** [comment]
based on our offline discussion, let's make sure this doesn't break minified mode

File: scvi/model/base/_utils.py [hunk 7]
```diff
@@ -14,13 +14,84 @@
...
+        **{
+            name: torch.index_select(
+                getattr(my_distribution, name), dim=dim, index=index
+            )
+            for name in my_distribution.arg_constraints.keys()
+        }
+    )
+
+
+def move_distribution(my_distribution, device):
```
Comment (adamgayoso) [inline scvi/model/base/_utils.py:37]: Let's use this

https://github.com/pyro-ppl/pyro/blob/9afb089f7f1f32b2bba5ce3064bee2769823285a/pyro/distributions/util.py#L159-L177

File: scvi/model/base/_utils.py [hunk 8]
```diff
@@ -14,13 +14,84 @@
...
+    """Utility function to move the parameter of a Pytorch distribution to a specific device."""
+    new_params = {
+        name: _move_data_to_device(getattr(my_distribution, name), device)
+        for name in my_distribution.arg_constraints.keys()
+    }
+    return my_distribution.__class__(**new_params)
+
+
+class DistributionsConcatenator:
```
Comment (adamgayoso) [inline scvi/model/base/_utils.py:46]: this should go in scvi.distributions

Comment (adamgayoso) [reply scvi/model/base/_utils.py:46]: shouldn't there be a way to find the batch dimension of the distribution? and make this more general?

Comment (adamgayoso) [reply scvi/model/base/_utils.py:46]: or are we constraining this to distributions with 1 batch dim

Comment (adamgayoso) [reply scvi/model/base/_utils.py:46]: we also need to test this

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/module/_totalvae.py [hunk 9]
```diff
@@ -683,7 +683,7 @@ def sample(self, tensors, n_samples=1):
 
     @torch.inference_mode()
     @auto_move_data
-    def marginal_ll(self, tensors, n_mc_samples):
+    def marginal_ll(self, tensors, n_mc_samples, observation_specific: bool = False):
```
[the remark this answers is not in the document]
Comment (PierreBoyeau) [reply scvi/module/_totalvae.py:689]: TOTALVAE's forward returns an error if you run it with `n_samples > 1` right now.
I can fix it, but this requires to change `TOTALVAE`'s inference and generative functions. Should I go for it?

File: scvi/model/base/_rnamixin.py [hunk 10]
```diff
@@ -30,6 +36,95 @@ def _get_transform_batch_gen_kwargs(self, batch):
...
+
+        log_px = self.get_marginal_ll(
+            adata,
+            indices=indices[anchor_cells],
+            observation_specific=True,
+            n_mc_samples=n_mc_samples,
+            n_mc_samples_per_pass=n_mc_samples_per_pass,
+        )
+        mask = torch.tensor(anchor_cells)
+        qz_anchor = subset_distribution(
```
Comment (adamgayoso) [inline scvi/model/base/_rnamixin.py:89]: pyro has a `MaskedDistribution` we should use and then can remove this util

File: scvi/distributions/_utils.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class DistributionsConcatenator:
+    """Utility class to concatenate Pytorch distributions and move them to cpu."""
+
+    def __init__(self):
+        self.storage = {}
+
+    def add_distributions(self, forward_outputs: dict):
+        """Add a dictionary of distributions to the concatenator."""
```
Comment (adamgayoso) [inline scvi/distributions/_utils.py:29]: need more docs to say that you can pass non dists in here

File: scvi/model/base/_rnamixin.py [hunk 12]
```diff
@@ -30,6 +36,95 @@ def _get_transform_batch_gen_kwargs(self, batch):
                 "Transforming batches is not implemented for this model."
             )
 
+    def get_importance_weights(
```
Comment (adamgayoso) [inline scvi/model/base/_rnamixin.py:39]: why should this be a public method?

the distributions argument is not very clear as well. I'm not sure how anyone would know what to feed here.

File: scvi/model/base/_rnamixin.py [hunk 13]
```diff
@@ -122,34 +225,63 @@ def get_normalized_expression(
...
+                exp_ = exp_[..., gene_mask]
+                exp_ *= scaling
+                per_batch_exprs.append(exp_[None].cpu())
+                if store_distributions:
+                    dist_store.add_distributions(
+                        {
+                            **inference_outputs,
+                            **generative_outputs,
+                        }
+                    )
```
Comment (adamgayoso) [inline scvi/model/base/_rnamixin.py:258]: This can be more clear. I don't like so much the filtering of distributions inside this object as we already know which two we want

File: scvi/distributions/_utils.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    )
+
+
+class DistributionsConcatenator:
+    """Utility class to concatenate Pytorch distributions and move them to cpu."""
+
+    def __init__(self):
+        self.storage = {}
+
+    def add_distributions(self, forward_outputs: dict):
```
Comment (adamgayoso) [inline scvi/distributions/_utils.py:28]:
```suggestion
    def store_distributions(self, forward_outputs: dict):
```

**Review by adamgayoso — COMMENTED** [review]
Still have a few important comments we should discuss

File: scvi/model/base/_rnamixin.py (hunk 10 above)
Comment (adamgayoso) [reply scvi/model/base/_rnamixin.py:89]: actually this won't work

File: scvi/model/base/_rnamixin.py [hunk 15]
```diff
@@ -30,6 +36,112 @@ def _get_transform_batch_gen_kwargs(self, batch):
...
+        subpopulation.
+
+        Parameters
+        ----------
+        adata
+            Data to use for computing importance weights.
+        indices
+            Indices of cells in adata to use.
+        distributions
+            Dictionary of distributions associated with `indices`.
```
Comment (adamgayoso) [inline scvi/model/base/_rnamixin.py:64]:
```suggestion
```

File: scvi/model/base/_rnamixin.py [hunk 16]
```diff
@@ -30,6 +36,112 @@ def _get_transform_batch_gen_kwargs(self, batch):
...
+            Number of Monte Carlo samples to use for each pass, by default 250
+
+        Returns
+        -------
+        importance_weights
+            Numpy array containing importance weights aligned with the provided `indices`.
+
+        Notes
+        -----
+            This method assumes a normal prior on the latent space.
```
Comment (adamgayoso) [inline scvi/model/base/_rnamixin.py:89]:
```suggestion
        This method assumes a normal prior on the latent space.
```

File: scvi/model/base/_rnamixin.py [hunk 17]
```diff
@@ -84,13 +201,19 @@ def get_normalized_expression(
         -------
         If `n_samples` > 1 and `return_mean` is False, then the shape is `(samples, cells, genes)`.
         Otherwise, shape is `(cells, genes)`. In this case, return type is :class:`~pandas.DataFrame` unless `return_numpy` is True.
+        This function has two modes of operation:
+        If `n_samples` is provided or if `return_mean` is True,
+        this method returns a 3d tensor of shape (n_samples, n_cells, n_genes).
+        Otherwise, the method expects `n_samples_overall` to be provided and returns a 2d tensor
+        of shape (n_samples_overall, n_genes).
```
Comment (adamgayoso) [inline scvi/model/base/_rnamixin.py:208]: can you clean this up with the text just above it?

File: scvi/distributions/_negative_binomial.py [hunk 18]
```diff
@@ -412,8 +413,8 @@ class ZeroInflatedNegativeBinomial(NegativeBinomial):
     arg_constraints = {
         "mu": constraints.greater_than_eq(0),
         "theta": constraints.greater_than_eq(0),
-        "zi_probs": constraints.half_open_interval(0.0, 1.0),
         "zi_logits": constraints.real,
+        "scale": constraints.greater_than_eq(0),
```
Comment (adamgayoso) [inline scvi/distributions/_negative_binomial.py:417]: does this force us to always provide scale?

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/distributions/_negative_binomial.py (hunk 18 above)
Comment (PierreBoyeau) [reply scvi/distributions/_negative_binomial.py:417]: I don't think so, `test_distributions.py` runs smoothly even though `scale` is not provided.

File: scvi/model/base/_rnamixin.py (hunk 17 above)
Comment (PierreBoyeau) [reply scvi/model/base/_rnamixin.py]: Done!

**Review by martinkim0 — APPROVED** [review]
LGTM

**Review by adamgayoso — APPROVED** [review]
LGTM, but @PierreBoyeau can you add a release note?
